File: VB/MultiLineButtonEdit.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Drawing
Imports System.Reflection
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraEditors.Registrator
Imports DevExpress.XtraEditors.Drawing
Imports DevExpress.XtraEditors.ViewInfo
Imports DevExpress.XtraEditors.Controls
Imports System.ComponentModel
Imports DevExpress.Utils
Imports DevExpress.Utils.Drawing
Imports System.Windows.Forms

Namespace MultilineButtonEditExample

	<UserRepositoryItem("RegisterMultiLineButtonEdit")> _
	Public Class RepositoryItemMultiLineButtonEdit
		Inherits RepositoryItemButtonEdit
		Private _acceptsTab As Boolean
		Private _acceptsReturn As Boolean
		Private _linesCount As Integer
		Private _scrollBars As ScrollBars
		Public Const MultiLineButtonEditName As String = "MultiLineButtonEdit"

		Shared Sub New()
			RegisterMultiLineButtonEdit()
		End Sub

		Public Sub New()
			Me._acceptsTab = False
			Me._acceptsReturn = True
			Me._linesCount = 0
			Me._scrollBars = ScrollBars.None
		End Sub

        <Browsable(False)> _
        Protected Overrides ReadOnly Property UseMaskBox() As Boolean
            Get
                Return False
            End Get
        End Property

        Public Overrides Property AutoHeight As Boolean
            Get
                Return False
            End Get
            Set(value As Boolean)
                MyBase.AutoHeight = value
            End Set
        End Property


		Public Overrides ReadOnly Property EditorTypeName() As String
			Get
				Return MultiLineButtonEditName
			End Get
		End Property

		Public Shared Sub RegisterMultiLineButtonEdit()
			Dim img As Image = Nothing
						Try
							img = CType(Bitmap.FromStream(System.Reflection.Assembly.GetAssembly(GetType(MultiLineButtonEdit)).GetManifestResourceStream("MultiLineButtonEdit.bmp")), Image)
					  Catch
					  End Try
			  EditorRegistrationInfo.Default.Editors.Add(New EditorClassInfo(MultiLineButtonEditName, GetType(MultiLineButtonEdit), GetType(RepositoryItemMultiLineButtonEdit), GetType(MultiLineButtonEditViewInfo), New ButtonEditPainter(), True, img, GetType(DevExpress.Accessibility.ButtonEditAccessible)))
		End Sub

		<DefaultValue(False)> _
		Public Property AcceptsTab() As Boolean
			Get
				Return _acceptsTab
			End Get
			Set(ByVal value As Boolean)
				If AcceptsTab = value Then
					Return
				End If
				_acceptsTab = value
				OnPropertiesChanged()
			End Set
		End Property

		<DefaultValue(True)> _
		Public Property AcceptsReturn() As Boolean
			Get
				Return _acceptsReturn
			End Get
			Set(ByVal value As Boolean)
				If AcceptsReturn = value Then
					Return
				End If
				_acceptsReturn = value
				OnPropertiesChanged()
			End Set
		End Property

		<Browsable(False)> _
		Public Property LinesCount() As Integer
			Get
				Return _linesCount
			End Get
			Set(ByVal value As Integer)
				If value < 1 Then
					value = 0
				End If
				If LinesCount = value Then
					Return
				End If
				_linesCount = value
				OnPropertiesChanged()
			End Set
		End Property

		Public Property ScrollBars() As ScrollBars
			Get
				Return _scrollBars
			End Get
			Set(ByVal value As ScrollBars)
				If ScrollBars = value Then
					Return
				End If
				_scrollBars = value
				OnPropertiesChanged()
			End Set
		End Property

		Public Overrides Sub Assign(ByVal item As RepositoryItem)
			Dim source As RepositoryItemMultiLineButtonEdit = TryCast(item, RepositoryItemMultiLineButtonEdit)
			BeginUpdate()
			Try
				MyBase.Assign(item)
				If source Is Nothing Then
					Return
				End If
				Me._acceptsReturn = source.AcceptsReturn
				Me._acceptsTab = source.AcceptsTab
				Me._scrollBars = source.ScrollBars
				Me._linesCount = source.LinesCount
			Finally
				EndUpdate()
			End Try
		End Sub

		Protected Overrides Function NeededKeysContains(ByVal key As Keys) As Boolean
			If AcceptsReturn AndAlso key = Keys.Enter Then
				Return True
			End If
			If AcceptsTab AndAlso key = Keys.Tab Then
				Return True
			End If
			If key = Keys.Up Then
				Return True
			End If
			If key = Keys.Down Then
				Return True
			End If
			Return MyBase.NeededKeysContains(key)
		End Function

		Protected Overrides Function IsNeededKeyCore(ByVal keyData As Keys) As Boolean
			If keyData = (Keys.Enter Or Keys.Control) Then
				Return False
			End If
			Dim res As Boolean = MyBase.IsNeededKeyCore(keyData)
			If res Then
				Return True
			End If
			If keyData = Keys.PageUp OrElse keyData = Keys.PageDown Then
				Return True
			End If
			Return False
		End Function
	End Class

	Public Class MultiLineButtonEditViewInfo
		Inherits ButtonEditViewInfo
		Implements IHeightAdaptable

		Public Sub New(ByVal item As RepositoryItem)
			MyBase.New(item)
		End Sub
		Public Shadows ReadOnly Property Item() As RepositoryItemMultiLineButtonEdit
			Get
				Return TryCast(MyBase.Item, RepositoryItemMultiLineButtonEdit)
			End Get
		End Property

		Private Shared defaultEditTextOptions_Renamed As TextOptions
		Private Shared ReadOnly Property DefaultEditTextOptions() As TextOptions
			Get
				If defaultEditTextOptions_Renamed Is Nothing Then
					defaultEditTextOptions_Renamed = New TextOptions(HorzAlignment.Near, VertAlignment.Top, WordWrap.Wrap, Trimming.None)
				End If
				Return defaultEditTextOptions_Renamed
			End Get
		End Property
		Public Overrides ReadOnly Property DefaultTextOptions() As TextOptions
			Get
				If OwnerEdit IsNot Nothing AndAlso OwnerEdit.InplaceType = InplaceType.Standalone Then
					If Item.ScrollBars = ScrollBars.Horizontal OrElse Item.ScrollBars = ScrollBars.Both Then
						Return New TextOptions(HorzAlignment.Near, VertAlignment.Top, WordWrap.NoWrap, Trimming.None)
					End If
				End If
				Return DefaultEditTextOptions
			End Get
		End Property

		Protected Overrides Sub CalcContentRect(ByVal bounds As Rectangle)
			MyBase.CalcContentRect(bounds)
			Me.fMaskBoxRect = ContentRect
			If Not(TypeOf BorderPainter Is EmptyBorderPainter) Then
				Me.fMaskBoxRect.Inflate(-1, -1)
			End If
		End Sub

		Private Function CalcHeight(ByVal cache As GraphicsCache, ByVal width As Integer) As Integer Implements IHeightAdaptable.CalcHeight
				Dim info As New BorderObjectInfoArgs(cache)
				info.Bounds = New Rectangle(0, 0, ContentRect.Width, 100)
			Dim textRect As Rectangle = BorderPainter.GetObjectClientRectangle(info)
			If Not(TypeOf BorderPainter Is EmptyBorderPainter) AndAlso Not(TypeOf BorderPainter Is InplaceBorderPainter) Then
				textRect.Inflate(-1, -1)
			End If
			Dim text As String = String.Empty
			If Item.LinesCount = 0 Then
				text = DisplayText
				If text IsNot Nothing AndAlso text.Length > 0 Then
					Dim lastChar As Char = text.Chars(text.Length - 1)
					If AscW(lastChar) = 13 OrElse AscW(lastChar) = 10 Then
						text &= "W"
					End If
				End If
			Else
				For i As Integer = 0 To Item.LinesCount - 1
					text &= (If(String.IsNullOrEmpty(text), "", Environment.NewLine)) & "W"
				Next i
			End If
			Dim height As Integer = CalcTextSizeCore(cache, text, textRect.Width).Height + 1
			Return (height + 100 - textRect.Bottom) + 1
		End Function
	End Class

	<ToolboxBitmapAttribute(GetType(MultiLineButtonEdit), "MultiLineButtonEdit.bmp"), ToolboxItem(True), Description("Custom ButtonEdit with multi line support.")> _
	Public Class MultiLineButtonEdit
		Inherits ButtonEdit
		Shared Sub New()
			RepositoryItemMultiLineButtonEdit.RegisterMultiLineButtonEdit()
		End Sub
		Public Sub New()
		End Sub
		Public Overrides ReadOnly Property EditorTypeName() As String
			Get
				Return RepositoryItemMultiLineButtonEdit.MultiLineButtonEditName
			End Get
		End Property

		<DesignerSerializationVisibility(DesignerSerializationVisibility.Content)> _
		Public Shadows ReadOnly Property Properties() As RepositoryItemMultiLineButtonEdit
			Get
				Return TryCast(MyBase.Properties, RepositoryItemMultiLineButtonEdit)
			End Get
		End Property
		Protected Overrides ReadOnly Property AcceptsReturn() As Boolean
			Get
				Return Properties.AcceptsReturn
			End Get
		End Property
		Protected Overrides ReadOnly Property AcceptsTab() As Boolean
			Get
				Return Properties.AcceptsTab
			End Get
		End Property

		Protected Overrides Sub UpdateMaskBoxProperties(ByVal always As Boolean)
			MyBase.UpdateMaskBoxProperties(always)
			If MaskBox Is Nothing Then
				Return
			End If
			If always OrElse (Not MaskBox.Multiline) Then
				MaskBox.Multiline = True
			End If
			If always OrElse (Not MaskBox.WordWrap) Then
				MaskBox.WordWrap = True
			End If
			If always OrElse Properties.AcceptsTab <> MaskBox.AcceptsTab Then
				MaskBox.AcceptsTab = Properties.AcceptsTab
			End If
			If always OrElse Properties.AcceptsReturn <> MaskBox.AcceptsReturn Then
				MaskBox.AcceptsReturn = Properties.AcceptsReturn
			End If
			If always OrElse Properties.ScrollBars <> MaskBox.ScrollBars Then
				MaskBox.ScrollBars = Properties.ScrollBars
			End If
		End Sub

		Protected Overrides Function IsInputKey(ByVal keyData As Keys) As Boolean
			Dim result As Boolean = MyBase.IsInputKey(keyData)
			If result Then
				Return True
			End If
			If Properties.AcceptsReturn AndAlso keyData = Keys.Enter Then
				Return True
			End If
			If Properties.AcceptsTab AndAlso keyData = Keys.Tab Then
				Return True
			End If
			Return result
		End Function
	End Class
End Namespace

File: CS/MultiLineButtonEdit.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Drawing;
using System.Reflection;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraEditors.Registrator;
using DevExpress.XtraEditors.Drawing;
using DevExpress.XtraEditors.ViewInfo;
using DevExpress.XtraEditors.Controls;
using System.ComponentModel;
using DevExpress.Utils;
using DevExpress.Utils.Drawing;
using System.Windows.Forms;

namespace MultilineButtonEditExample {

    [UserRepositoryItem("RegisterMultiLineButtonEdit")]
    public class RepositoryItemMultiLineButtonEdit : RepositoryItemButtonEdit {
        bool _acceptsTab;
        bool _acceptsReturn;
        int _linesCount;
        ScrollBars _scrollBars;
        public const string MultiLineButtonEditName = "MultiLineButtonEdit";

        static RepositoryItemMultiLineButtonEdit() { RegisterMultiLineButtonEdit(); }

        public RepositoryItemMultiLineButtonEdit() {
            this._acceptsTab = false;
            this._acceptsReturn = true;
            this._linesCount = 0;
            this._scrollBars = ScrollBars.None;
        }

        [Browsable(false)]
        protected override bool UseMaskBox { get { return false; } }
        [Browsable(false)]
        public override bool AutoHeight { get { return false; } }

        public override string EditorTypeName { get { return MultiLineButtonEditName; } }

        public static void RegisterMultiLineButtonEdit() {
            Image img = null;
                        try {
                            img = (Image)Bitmap.FromStream(Assembly.GetAssembly(typeof(MultiLineButtonEdit)).GetManifestResourceStream("MultilineButtonEditExample.MultiLineButtonEdit.bmp"));
                      } catch { }
              EditorRegistrationInfo.Default.Editors.Add(new EditorClassInfo(MultiLineButtonEditName,
              typeof(MultiLineButtonEdit),
              typeof(RepositoryItemMultiLineButtonEdit),
              typeof(MultiLineButtonEditViewInfo),
              new ButtonEditPainter(),
              true,
              img,
              typeof(DevExpress.Accessibility.ButtonEditAccessible)));
        }

        [DefaultValue(false)]
        public bool AcceptsTab {
            get { return _acceptsTab; }
            set {
                if(AcceptsTab == value) return;
                _acceptsTab = value;
                OnPropertiesChanged();
            }
        }

        [DefaultValue(true)]
        public bool AcceptsReturn {
            get { return _acceptsReturn; }
            set {
                if(AcceptsReturn == value) return;
                _acceptsReturn = value;
                OnPropertiesChanged();
            }
        }
        
        [Browsable(false)]
        public int LinesCount {
            get { return _linesCount; }
            set {
                if(value < 1) value = 0;
                if(LinesCount == value) return;
                _linesCount = value;
                OnPropertiesChanged();
            }
        }

        public ScrollBars ScrollBars {
            get { return _scrollBars; }
            set {
                if(ScrollBars == value) return;
                _scrollBars = value;
                OnPropertiesChanged();
            }
        }

        public override void Assign(RepositoryItem item) {
            RepositoryItemMultiLineButtonEdit source = item as RepositoryItemMultiLineButtonEdit;
            BeginUpdate();
            try {
                base.Assign(item);
                if(source == null) return;
                this._acceptsReturn = source.AcceptsReturn;
                this._acceptsTab = source.AcceptsTab;
                this._scrollBars = source.ScrollBars;
                this._linesCount = source.LinesCount;
            } finally {
                EndUpdate();
            }
        }

        protected override bool NeededKeysContains(Keys key) {
            if(AcceptsReturn && key == Keys.Enter)
                return true;
            if(AcceptsTab && key == Keys.Tab)
                return true;
            if(key == Keys.Up)
                return true;
            if(key == Keys.Down)
                return true;
            return base.NeededKeysContains(key);
        }

        protected override bool IsNeededKeyCore(Keys keyData) {
            if(keyData == (Keys.Enter | Keys.Control)) return false;
            bool res = base.IsNeededKeyCore(keyData);
            if(res) return true;
            if(keyData == Keys.PageUp || keyData == Keys.PageDown) return true;
            return false;
        }
    }

    public class MultiLineButtonEditViewInfo : ButtonEditViewInfo, IHeightAdaptable {

        public MultiLineButtonEditViewInfo(RepositoryItem item) : base(item) { }
        public new RepositoryItemMultiLineButtonEdit Item { get { return base.Item as RepositoryItemMultiLineButtonEdit; } }

        static TextOptions defaultEditTextOptions;
        static TextOptions DefaultEditTextOptions {
            get {
                if(defaultEditTextOptions == null) {
                    defaultEditTextOptions = new TextOptions(HorzAlignment.Near, VertAlignment.Top, WordWrap.Wrap, Trimming.None);
                }
                return defaultEditTextOptions;
            }
        }
        public override TextOptions DefaultTextOptions {
            get {
                if(OwnerEdit != null && OwnerEdit.InplaceType == InplaceType.Standalone) {
                    if(Item.ScrollBars == ScrollBars.Horizontal || Item.ScrollBars == ScrollBars.Both) {
                        return new TextOptions(HorzAlignment.Near, VertAlignment.Top, WordWrap.NoWrap, Trimming.None);
                    }
                }
                return DefaultEditTextOptions;
            }
        }

        protected override void CalcContentRect(Rectangle bounds) {
            base.CalcContentRect(bounds);
            this.fMaskBoxRect = ContentRect;
            if(!(BorderPainter is EmptyBorderPainter))
                this.fMaskBoxRect.Inflate(-1, -1);
        }

        int IHeightAdaptable.CalcHeight(GraphicsCache cache, int width) {
                BorderObjectInfoArgs info = new BorderObjectInfoArgs(cache);
                info.Bounds = new Rectangle(0, 0, ContentRect.Width, 100);
            Rectangle textRect = BorderPainter.GetObjectClientRectangle(info);
            if(!(BorderPainter is EmptyBorderPainter) && !(BorderPainter is InplaceBorderPainter))
                textRect.Inflate(-1, -1);
            string text = string.Empty;
            if(Item.LinesCount == 0) {
                text = DisplayText;
                if(text != null && text.Length > 0) {
                    char lastChar = text[text.Length - 1];
                    if(lastChar == 13 || lastChar == 10) text += "W";
                }
            } else {
                for(int i = 0; i < Item.LinesCount; i++)
                    text += (string.IsNullOrEmpty(text) ? "" : Environment.NewLine) + "W";
            }
            int height = CalcTextSizeCore(cache, text, textRect.Width).Height + 1;
            return (height + 100 - textRect.Bottom) + 1;
        }
    }

    [ToolboxBitmapAttribute(typeof(MultiLineButtonEdit), "MultiLineButtonEdit.bmp")]
    [ToolboxItem(true)]
    [Description("Custom ButtonEdit with multi line support.")]
    public class MultiLineButtonEdit : ButtonEdit {
        static MultiLineButtonEdit() { RepositoryItemMultiLineButtonEdit.RegisterMultiLineButtonEdit(); }
        public MultiLineButtonEdit() { }
        public override string EditorTypeName { get { return RepositoryItemMultiLineButtonEdit.MultiLineButtonEditName; } }

        [DesignerSerializationVisibility(DesignerSerializationVisibility.Content)]
        public new RepositoryItemMultiLineButtonEdit Properties {
            get { return base.Properties as RepositoryItemMultiLineButtonEdit; }
        }
        protected override bool AcceptsReturn { get { return Properties.AcceptsReturn; } }
        protected override bool AcceptsTab { get { return Properties.AcceptsTab; } }

        protected override void UpdateMaskBoxProperties(bool always) {
            base.UpdateMaskBoxProperties(always);
            if(MaskBox == null) return;
            if(always || !MaskBox.Multiline) MaskBox.Multiline = true;
            if(always || !MaskBox.WordWrap) MaskBox.WordWrap = true;
            if(always || Properties.AcceptsTab != MaskBox.AcceptsTab) MaskBox.AcceptsTab = Properties.AcceptsTab;
            if(always || Properties.AcceptsReturn != MaskBox.AcceptsReturn) MaskBox.AcceptsReturn = Properties.AcceptsReturn;
            if(always || Properties.ScrollBars != MaskBox.ScrollBars) MaskBox.ScrollBars = Properties.ScrollBars;
        }

        protected override bool IsInputKey(Keys keyData) {
            bool result = base.IsInputKey(keyData);
            if(result) return true;
            if(Properties.AcceptsReturn && keyData == Keys.Enter) return true;
            if(Properties.AcceptsTab && keyData == Keys.Tab) return true;
            return result;
        }
    }
}
